### PR TITLE
Add UI testing

### PR DESCRIPTION
**Pull request by jmandel**
This PR adds UI testing through karma, both locally and in Circle CI.

In addition to the CircleCI config, you'll see tweaks to some of the typescript and karma config files, which represent the result of looking carefully through the differences between this project and a brand-new repo created via `ng new` — since tests were failing in this project, while working in the `ng new` project. The good news is that it's almost as many deletions as additions :-)

The current approach uses a Circle CI "workflow" to run api and ui tests in parallel, in two separate/isolated environments.

### CODE REVIEW

File: ui/package.json [hunk 1]
```diff
@@ -22,7 +22,7 @@
     "@angular/router": "^4.0.0",
     "core-js": "^2.4.1",
     "rxjs": "^5.1.0",
-    "zone.js": "^0.8.4"
+    "zone.js": "0.8.12"
```
Comment (jmandel) [inline ui/package.json]: This is to deal with https://github.com/angular/zone.js/issues/832

File: ui/tslint.json [hunk 2]
```diff
@@ -25,13 +26,19 @@
     "label-position": true,
     "max-line-length": [
       true,
-      100
+      140
```
Comment (markfickett) [inline ui/tslint.json]: I'd like to stick with 100.

File: ui/src/tsconfig.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (markfickett) [inline ui/src/tsconfig.json]: Why is this file needed? I think there are already two tsconfig.json files under `ui/`, though one might be tsconfig.app.json or something.

File: ui/src/environments/environment.prod.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-export const environment = {
+export const environment: object = {
```
Comment (markfickett) [inline ui/src/environments/environment.prod.ts]: What does this do?

Should this apply to the other environment files?

File: ui/src/app/views/app/component.spec.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import { TestBed, async } from '@angular/core/testing';
```
Comment (markfickett) [inline ui/src/app/views/app/component.spec.ts]: Yay, thanks for setting up tests!

File: .circleci/config.yml [hunk 6]
```diff
@@ -22,3 +22,30 @@ jobs:
...
+      - run:
+          working_directory: ~/workbench/ui
+          command: npm install
+      - save_cache:
+          paths:
+            - ~/workbench/ui/node_modules
+          key: ui-cache-{{ checksum "~/workbench/ui/package.json" }}
+      - run:
+          working_directory: ~/workbench/ui
+          command: npm run test -- --no-watch
```
Comment (markfickett) [inline .circleci/config.yml]: Is it the --no-watch that keeps us from trying to open browsers?

File: .circleci/config.yml [hunk 7]
```diff
@@ -22,3 +22,30 @@ jobs:
             - ~/.m2
             - ~/workbench/api/build/exploded-api/WEB-INF/lib/
           key: api-cache-{{ checksum "~/workbench/api/build.gradle" }}
+  build_ui:
```
Comment (markfickett) [inline .circleci/config.yml]: How about "build_and_test_ui"?

File: ui/tslint.json (hunk 2 above)
Comment (jmandel) [inline ui/tslint.json]: I'll set it back down (this came in as a default among a bunch of settings from the output of `ng new`).

File: ui/src/tsconfig.json (hunk 3 above)
Comment (jmandel) [inline ui/src/tsconfig.json]: I haven't explored deeply -- this is just the default project structure that `ng new` creates (and it works).

File: ui/src/environments/environment.prod.ts (hunk 4 above)
Comment (jmandel) [inline ui/src/environments/environment.prod.ts]: This resolved a typing error I was getting on `ng build`.

File: .circleci/config.yml (hunk 6 above)
Comment (jmandel) [inline .circleci/config.yml]: karma (the thing that runs when `ng test` is invoked, as via `npm run test`) always tries to open browsers. The default behavior if you just do `ng test` (or `npm run test`) is to open a browser and keep it open, watching your files and re-running tests as you edit your code. Which is nice when you're debugging locally, but not great in CI (because it never finishes). Passing `--no-watch` runs the tests once and then closes out the browser (succeed or fail).

File: .circleci/config.yml (hunk 7 above)
Comment (jmandel) [inline .circleci/config.yml]: Done.
